graphite/graphite.go

```diff
@@ -0,0 +1,330 @@
+			return err
+		}
+		defer func() { _ = l.Close() }()
+
+		s.wg.Add(1)
+		go s.serveUDP(l, done)
+	}
+
+	return nil
+}
+
+// serveTCP handles incoming TCP connection requests.
+func (s *Server) serveTCP(l *net.TCPListener, done chan struct{}) {
+	defer s.wg.Done()
+
+	// Listen for server close.
+	go func() {
+		<-done
+		l.Close()
+	}()
+
+	// Listen for new TCP connections.
+	for {
+		c, err := l.Accept()
+		if err != nil {
+			// TODO(benbjohnson): Check for connection closed.
+			log.Error("graphite.Server: Accept: ", err)
+			continue
+		}
+
+		s.wg.Add(1)
+		go s.handleTCPConn(c)
+	}
+}
+
+func (s *Server) handleTCPConn(conn net.Conn) {
+	defer conn.Close()
+	defer s.wg.Done()
+
+	reader := bufio.NewReader(conn)
+	for {
+		err := s.handleMessage(reader)
+		if err != nil {
+			if io.EOF == err {
+				log.Debug("graphite.Server: Client closed graphite connection")
+				return
+			}
+			log.Error("graphite.Server:", err)
+		}
+	}
+}
+
+// serveUDP handles incoming UDP messages.
+func (s *Server) serveUDP(conn *net.UDPConn, done chan struct{}) {
+	defer s.wg.Done()
+
+	// Listen for server close.
+	go func() {
+		<-done
+		conn.Close()
+	}()
+
+	buf := make([]byte, 65536)
+	for {
+		// Read from connection.
+		n, _, err := conn.ReadFromUDP(buf)
+		if err == io.EOF {
+			return
+		} else if err != nil {
+			log.Warn("Server: Error when reading from UDP connection %s", err.Error())
+		}
+
+		// Read in data in a separate goroutine.
+		s.wg.Add(1)
+		go s.handleUDPMessage(string(buf[:n]))
+	}
+}
+
+// handleUDPMessage splits a UDP packet by newlines and processes each message.
+func (s *Server) handleUDPMessage(msg string) {
+	defer s.wg.Done()
+	for _, metric := range strings.Split(msg, "\n") {
+		s.handleMessage(bufio.NewReader(strings.NewReader(metric + "\n")))
+	}
+}
+
+// Close shuts down the server's listeners.
+func (s *Server) Close() error {
+	// Notify other goroutines of shutdown.
+	s.mu.Lock()
+	if s.done == nil {
+		s.mu.Unlock()
+		return ErrServerClosed
+	}
+	close(s.done)
+	s.done = nil
+	s.mu.Unlock()
+
+	// Wait for all goroutines to shutdown.
+	s.wg.Wait()
+	return nil
+}
+
+// handleMessage decodes a graphite message from the reader and sends it to the
+// committer goroutine.
+func (s *Server) handleMessage(r *bufio.Reader) error {
+	// Decode graphic metric.
+	m, err := DecodeMetric(r)
+	if err != nil {
+		return err
+	}
+
+	// Convert metric to a field value.
+	var values = make(map[string]interface{})
+	if m.IsInt {
+		values[m.Name] = &m.IntegerValue
+	} else {
+		values[m.Name] = &m.FloatValue
+	}
+
+	retentionPolicy, err := s.Server.DefaultRetentionPolicy(s.Database)
+
+	if err != nil {
+		return fmt.Errorf("error looking up default database retention policy: %s", err)
+	}
+
+	if err := s.Server.WriteSeries(
+		s.Database,
+		retentionPolicy.Name,
+		m.Name,
+		m.Tags,
+		m.Timestamp,
+		values,
+	); err != nil {
+		return fmt.Errorf("write series data: %s", err)
+	}
+	return nil
+}
+
+type Metric struct {
```

Comment:
Should `Metric` be an exported type? If so I think we should have some standardization on this between the HTTP API, Graphite, and other input plugins.